src/CloudNative.CloudEvents.Protobuf/ProtobufEventFormatter.cs

```diff
@@ -0,0 +1,412 @@
+                };
+            }
+
+            DecodeStructuredModeData(proto, cloudEvent);
+
+            return Validation.CheckCloudEventArgument(cloudEvent, paramName);
+        }
+
+        /// <summary>
+        /// Decodes the "data" property provided within a structured-mode message,
+        /// populating the <see cref="CloudEvents.CloudEvent.Data"/> property accordingly.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// This implementation simply converts binary data to a byte array, leaves proto data
+        /// as an <see cref="Google.Protobuf.WellKnownTypes.Any"/>, and converts text data to a string.
+        /// </para>
+        /// <para>
+        /// Override this method to provide more specialized conversions, such as to use <see cref="ByteString"/>
+        /// instead of a byte array, or to "unwrap" the proto data to generated code.
+        /// </para>
+        /// </remarks>
+        /// <param name="proto">The protobuf representation of the CloudEvent. Will not be null.</param>
+        /// <param name="cloudEvent">The event being decoded. This should not be modified except to
+        /// populate the <see cref="CloudEvents.CloudEvent.Data"/> property, but may be used to provide extra
+        /// information such as the data content type. Will not be null.</param>
+        /// <returns>The data to populate in the <see cref="CloudEvents.CloudEvent.Data"/> property.</returns>
+        protected virtual void DecodeStructuredModeData(V1.CloudEvent proto, CloudEvent cloudEvent) =>
+            cloudEvent.Data = proto.DataCase switch
+            {
+                DataOneofCase.BinaryData => proto.BinaryData.ToByteArray(),
+                DataOneofCase.ProtoData => proto.ProtoData,
+                DataOneofCase.TextData => proto.TextData,
+                DataOneofCase.None => null,
+                // Note: impossible to cover in tests
+                _ => throw new ArgumentException($"Unhandled protobuf data case: {proto.DataCase}")
+            };
+
+        /// <summary>
+        /// Encodes structured (or batch) mode data within a CloudEvent, storing it in the specified <see cref="CloudEvents.CloudEvent"/>.
+        /// </summary>
+        /// <param name="cloudEvent">The CloudEvent being encoded, which will have a non-null value for
+        /// its <see cref="CloudEvents.CloudEvent.Data"/> property.</param>
+        /// <param name="proto">The protobuf representation of the CloudEvent, which will be non-null.</param>
+        protected virtual void EncodeStructuredModeData(CloudEvent cloudEvent, V1.CloudEvent proto)
+        {
+            switch (cloudEvent.Data)
+            {
+                case IMessage message:
+                    proto.ProtoData = message is Any any ? any : Any.Pack(message, TypeUrlPrefix);
+                    break;
+                case string text:
```

Comment:
Does content type matter at all here?

Reply:
Nope. The structured mode format captures that it's text data, so we're able to retrieve it as a string later on regardless of content type.